kernel/vfs/src/node/mod.rs

```diff
@@ -0,0 +1,50 @@
+mod directory;
+mod file;
+
+pub use directory::Directory;
+pub use file::{File, SeekFrom};
+
+use crate::{file_system, Path};
+use alloc::{string::String, sync::Arc};
+
+/// A file system node.
+///
+/// Types implementing this trait should also implement either `File` or
+/// `Directory`.
+pub trait Node {
+    /// Returns the node's name.
+    fn name(&self) -> String;
+
+    /// Sets the node's name.
+    fn set_name(&self, name: String);
+
+    /// Returns the node's parent.
+    ///
+    /// A return value of [`None`] either indicates `self` is a root directory,
+    /// or its parent has been removed.
+    fn parent(&self) -> Option<Arc<dyn Directory>>;
+
+    /// Returns the specific node kind.
+    fn as_specific(self: Arc<Self>) -> NodeKind;
+}
+
+/// A specific file system node kind.
+pub enum NodeKind {
+    Directory(Arc<dyn Directory>),
+    File(Arc<dyn File>),
+}
+
+/// Gets the node at the `path`.
+///
+/// The path consists of a file system id, followed by a colon, followed by the
+/// path. For example, `tmp:/a/b` or `nvme:/foo/bar`.
```

Comment:
The use of colon here could be debated, I'm not particularly against it though.

Reply:
As I said in the PR, I just copied Redox but this is definitely up for debate if anyone has any better ideas.

Reply:
If we used the first path node name as fs name (`/root-fs/usr/bin/gcc` or `/usb-drive-2/photos/camping-trip-2016` for instance), that would ease filesystem enumeration but could also result in confusement and vulnerabilities.

Reply:
i'd probably prefer to have a more Unix-style set of mount points, but for simplicity's sake it doesn't have to be arbitrary. We could have all FSes mounted at `/` or `/mnt/`, for example.